src/lib.rs

```diff
@@ -1472,28 +1473,52 @@ pub mod __rt {
         }
 
         pub fn borrow(&self) -> Ref<T> {
-            unsafe {
-                if self.borrow.get() == usize::max_value() {
-                    borrow_fail();
-                }
-                self.borrow.set(self.borrow.get() + 1);
-                Ref {
-                    value: &*self.value.get(),
-                    borrow: &self.borrow,
+            let mut borrow = self.borrow.load(Ordering::SeqCst);
+            loop {
+                if borrow == usize::max_value() {
+                    borrow_fail()
                 }
+                match self.borrow.compare_exchange_weak(
+                    borrow,
+                    borrow + 1,
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                ) {
+                    Ok(_) => unsafe {
+                        return Ref {
+                            value: &*self.value.get(),
+                            borrow: &self.borrow,
+                        };
+                    },
+                    Err(new_borrow) => {
+                        borrow = new_borrow;
+                    }
+                };
             }
         }
 
         pub fn borrow_mut(&self) -> RefMut<T> {
-            unsafe {
-                if self.borrow.get() != 0 {
-                    borrow_fail();
-                }
-                self.borrow.set(usize::max_value());
-                RefMut {
-                    value: &mut *self.value.get(),
-                    borrow: &self.borrow,
+            let mut borrow = self.borrow.load(Ordering::SeqCst);
+            loop {
+                if borrow > 0 {
+                    borrow_fail()
                 }
+                match self.borrow.compare_exchange_weak(
+                    borrow,
+                    usize::max_value(),
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                ) {
+                    Ok(_) => unsafe {
+                        return RefMut {
+                            value: &mut *self.value.get(),
+                            borrow: &self.borrow,
+                        };
+                    },
+                    Err(new_borrow) => {
+                        borrow = new_borrow;
+                    }
+                };
             }
```

Comment:
Using a `compare_exchange()` here should do the trick without a `loop` and `load()`.
Also we shouldn't use the `weak` version of `compare_exchange()` unless we are fine with spurious fails, which we aren't, or am I missing something here?

Reply:
SGTM! We can `self.borrow.compare_exchange(0, usize::max_value())` and `borrow_fail()` if the `compare_exchange()` failed.